### DISCUSSION TITLE
AltDSS: adding PVs

### DISCUSSION THREAD

**Discussion by imperanatus** [dss-extensions/dss-extensions#38] [General]
It's my first time with OpenDSS and I've started with altDSS package. My question is: how do I select a set of buses and connect PV systems to them? There is a direct command to this or do I need to use text: for example altdss("New something")?

**Comment by PMeira** [comment]
(I'll assume you mean the Python package.)

> or do I need to use text: for example altdss("New something")?

That's an option. For the classic use of OpenDSS, people tend to generate the command strings and feed the engine. If you choose to do that, and are  working with large-scale circuits, using a big string with all commands is the fastest alternative.
 
> There is a direct command to this

Yes:

- for creating a single element, [altdss.PVSystem.new()](https://dss-extensions.org/AltDSS-Python/apidocs/altdss/altdss.PVSystem.html#altdss.PVSystem.IPVSystem.new)
- for creating a batch of elements, [altdss.PVSystem.batch_new()](https://dss-extensions.org/AltDSS-Python/apidocs/altdss/altdss.PVSystem.html#altdss.PVSystem.IPVSystem.batch_new)

If you're using an editor/IDE with good type hint support, it should autocomplete and suggest the properties. For a reference to the properties, see https://dss-extensions.org/dss-format/PVSystem.html (there's a link for the official OpenDSS docs and example .DSS scripts).

For the time being, some of the tests can help getting an idea of how it works:

- https://github.com/dss-extensions/AltDSS-Python/blob/main/tests/test_obj.py
- https://github.com/dss-extensions/AltDSS-Python/blob/main/tests/test_batch.py (this one is using the case insensitive option, so beware of that if you copy something)

**PVSystem has a lot of parameters** (if you don't really need those, you might want to use a generator or negative load instead), so I won't dare to write an example just before bed time, but just to illustrate a simplified command (without setting up efficiency and irradiance curves, maybe attach InvControl, etc.) could be something like:

```
pv_buses = ['bus1.1.2.3', 'bus2.1.2.3', 'bus3.1.2.3'] # select the bus connections somehow
pvs = altdss.PVSystem.batch_new(count=len(pv_buses), Phases=3, Bus1=pv_buses, kV=4.16, kVA=5)
```

For batches, most properties accept both scalars and arrays. So, if you're adding a list of uniform PVs, you can just pass a single value.

Since you mentioned you're new, remember to solve the system after the changes before trying to get the updated power flow state.

A few colleagues are writing some examples, but it should take a few more weeks before we publish them. Let me know if this was enough to get you started, otherwise I can try to write a simple PV example tomorrow with the whole setup.

**Comment by imperanatus** [comment]
Thank you, Paulo. I really liked this approach. It'll be very valuable and efficient for my studies. For now that's it. Other question. I'm not very familiar with forums in GitHub. So, Is this channel the most appropriate place to ask basic questions about the AltDss package?

**Reply by PMeira** [reply]
> I really liked this approach. It'll be very valuable and efficient for my studies. 

That's what we hope! There are some rough corners which will be handled in future versions, and the lack of docs, but I think we have a solid base now.

> So, Is this channel the most appropriate place to ask basic questions about the AltDss package?

Yes, it's just that there are new packages named AltDSS too, for other programming languages (Rust and Go).

**Reply by PMeira** [reply]
By the way, if you need to debug some specific case created from the AltDSS functions, you could save the circuit and load it on the official OpenDSS, etc.

I just noticed we forgot to add the new function in AltDSS, but you can use it from ODD.py or DSS-Python, `altdss.to_opendssdirect().ActiveCircuit.Save` or `altdss.to_dss_python().ActiveCircuit.Save`.  The new function provides more control of the output, and there are some examples in https://dss-extensions.org/OpenDSSDirect.py/notebooks/Saving.html

**Reply by imperanatus** [reply]
Nice Paulo. I really appreciate how altDSS-python works with LV circuits. I'm going to explore more this tool.
